src/serde/de.rs

```diff
@@ -0,0 +1,931 @@
+use crate::data_source::IonDataSource;
+use crate::result::IonFailure;
+use crate::serde::decimal::TUNNELED_DECIMAL_TYPE_NAME;
+use crate::serde::timestamp::TUNNELED_TIMESTAMP_TYPE_NAME;
+use crate::{
+    Decimal, IonError, IonReader, IonResult, IonType, ReaderBuilder, StreamItem, Symbol, Timestamp,
+};
+use serde::de;
+use serde::de::{DeserializeSeed, EnumAccess, MapAccess, SeqAccess, Visitor};
+use serde::Deserialize;
+use std::borrow::Cow;
+use std::iter::FusedIterator;
+use std::marker::PhantomData;
+
+/// Generic method that can deserialize an object from any given object
+/// that implements `ToIonDataSource`. It also enables `ION_BINARY`
+/// to ensure that `Timestamp` objects are correctly deserialized.
+pub fn from_ion<'a, T, S>(s: S) -> IonResult<T>
+where
+    T: Deserialize<'a>,
+    S: IonDataSource,
+{
+    let mut deserializer = Deserializer {
+        reader: ReaderBuilder::new().build(s)?,
+    };
+
+    if StreamItem::Nothing == deserializer.reader.current() {
+        // We're not on a value. Advance the reader.
+        deserializer.reader.next()?;
+    }
+
+    if StreamItem::Nothing == deserializer.reader.current() {
+        // Advancing the reader did nothing; this input contains no values.
+        return IonResult::decoding_error(
+            "The input for deserialization doesn't contain any values",
+        );
+    }
+
+    let result = T::deserialize(&mut deserializer)?;
+    Ok(result)
+}
+
+/// The deserializer for Ion, it doesn't care about
+/// whether the reader is reading binary or text representation.
+#[derive(Debug)]
+pub struct Deserializer<R> {
+    pub(crate) reader: R,
+}
+
+impl<'de, 'a, R> de::Deserializer<'de> for &'a mut Deserializer<R>
+where
+    R: IonReader<Symbol = Symbol, Item = StreamItem>,
+{
+    type Error = IonError;
+
+    fn deserialize_any<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: Visitor<'de>,
+    {
+        if let Some(ion_type) = self.reader.ion_type() {
+            match ion_type {
+                IonType::Null => self.deserialize_unit(visitor),
+                IonType::Blob | IonType::Clob => self.deserialize_bytes(visitor),
+                IonType::String | IonType::Symbol => self.deserialize_string(visitor),
+                IonType::Float => self.deserialize_f64(visitor),
+                IonType::Int => self.deserialize_i64(visitor),
+                IonType::Decimal => {
+                    self.deserialize_newtype_struct(TUNNELED_DECIMAL_TYPE_NAME, visitor)
+                }
+                IonType::Bool => self.deserialize_bool(visitor),
+                IonType::List => self.deserialize_seq(visitor),
+                IonType::Struct => self.deserialize_struct("", &[], visitor),
+                IonType::Timestamp => {
+                    self.deserialize_newtype_struct(TUNNELED_TIMESTAMP_TYPE_NAME, visitor)
+                }
+                _ => IonResult::decoding_error("unexpected ion type"),
```

Comment:
I can add deserialize_seq for SExp but we don't support serialization for SExp. This might be confusing for users which is why I wanted to keep SExp as not supported for both serialization and deserialization. Thoughts?

Reply:
I am not sure that we should deserialize lists and sexps as if they are interchangeable. I am sure that we should eventually have support for deserialization _and_ serialization of s-expressions. (Possibly configurable using `#[ion(sexp)]` or something like that on a struct or field.) I think that the configuration for serializing as a sexp might affect our decisions about how to deserialize a sexp. Therefore, I am in favor of punting both serialization and deserialization of sexps. Let's create an issue for it, and just make sure it gets resolved before we remove the "experimental" status of this feature.

Reply:
https://github.com/amazon-ion/ion-rust/issues/646